packages/components/carousel/src/carousel.vue

```diff
@@ -0,0 +1,324 @@
+  if (index < 0) {
+    activeIndex.value = props.loop ? length - 1 : 0
+  } else if (index >= length) {
+    activeIndex.value = props.loop ? 0 : length - 1
+  } else {
+    activeIndex.value = index
+  }
+  if (oldIndex === activeIndex.value) {
+    resetItemPosition(oldIndex)
+  }
+}
+
+function resetItemPosition(oldIndex) {
+  items.value.forEach((item, index) => {
+    item.translateItem(index, activeIndex.value, oldIndex)
+  })
+}
+
+function addItem(item) {
+  items.value.push(item)
+}
+
+function removeItem(uid) {
+  const index = items.value.findIndex((item) => item.uid === uid)
+  if (index !== -1) {
+    items.value.splice(index, 1)
+    if (activeIndex.value === index) next()
+  }
+}
+
+function itemInStage(item, index) {
+  const length = items.value.length
+  if (
+    (index === length - 1 && item.inStage && items.value[0].active) ||
+    (item.inStage && items.value[index + 1] && items.value[index + 1].active)
+  ) {
+    return 'left'
+  } else if (
+    (index === 0 && item.inStage && items.value[length - 1].active) ||
+    (item.inStage && items.value[index - 1] && items.value[index - 1].active)
+  ) {
+    return 'right'
+  }
+  return false
+}
+
+function handleMouseEnter() {
+  hover.value = true
+  if (props.pauseOnHover) {
+    pauseTimer()
+  }
+}
+
+function handleMouseLeave() {
+  hover.value = false
+  startTimer()
+}
+
+function handleButtonEnter(arrow) {
+  if (props.direction === 'vertical') return
+  items.value.forEach((item, index) => {
+    if (arrow === itemInStage(item, index)) {
+      item.hover = true
+    }
+  })
+}
+
+function handleButtonLeave() {
+  if (props.direction === 'vertical') return
+  items.value.forEach((item) => {
+    item.hover = false
+  })
+}
+
+function handleIndicatorClick(index) {
+  activeIndex.value = index
+}
+
+function handleIndicatorHover(index) {
+  if (props.trigger === 'hover' && index !== activeIndex.value) {
+    activeIndex.value = index
+  }
+}
+
+function prev() {
+  setActiveItem(activeIndex.value - 1)
+}
+
+function next() {
+  setActiveItem(activeIndex.value + 1)
+}
+
+// watch
+watch(
+  () => activeIndex.value,
+  (current, prev) => {
+    resetItemPosition(prev)
+    if (prev > -1) {
+      emit('change', current, prev)
+    }
+  }
+)
+watch(
+  () => props.autoplay,
+  (current) => {
+    current ? startTimer() : pauseTimer()
+  }
+)
+watch(
+  () => props.loop,
+  () => {
+    setActiveItem(activeIndex.value)
+  }
+)
+
+// lifecycle
+onMounted(() => {
+  nextTick(() => {
+    const resizeObserver = useResizeObserver(root.value, resetItemPosition)
+    if (props.initialIndex < items.value.length && props.initialIndex >= 0) {
+      activeIndex.value = props.initialIndex
+    }
+    startTimer()
+
+    onBeforeUnmount(() => {
+      if (root.value) resizeObserver.stop()
+      pauseTimer()
+    })
+  })
+})
+
+// provide
+provide(carouselContextKey, {
+  root,
+  direction: props.direction,
+  type: props.type,
+  items,
+  loop: props.loop,
```

Comment:
These action removes the reactivity of `direction`, `type`, `loop`. We can do something like:


```suggestion
  direction: toRef(props, 'direction'),
  type: toRef(props, 'type'),
  items,
  loop: toRef(props, 'loop'),
```